pkg/apis/core/validation/validation.go

```diff
@@ -3327,6 +3327,13 @@ func ValidatePodSpec(spec *core.PodSpec, podMeta *metav1.ObjectMeta, fldPath *fi
 		}
 	}
 
+	if spec.TerminationGracePeriodSeconds != nil {
```

Comment:
This needs to deal with invalid existing persisted data. As-is, this would prevent any existing invalid pods from being deleted.

Reply:
something like this is needed to avoid invalidating existing data:

```diff
diff --git a/pkg/api/pod/util.go b/pkg/api/pod/util.go
index 27398b53fc4..0d28a5dbde3 100644
--- a/pkg/api/pod/util.go
+++ b/pkg/api/pod/util.go
@@ -319,6 +319,10 @@ func usesHugePagesInProjectedEnv(item api.Container) bool {
 	return false
 }
 
+func hasNegativeTerminationGracePeriodSeconds(podSpec *api.PodSpec) bool {
+	return podSpec.TerminationGracePeriodSeconds != nil && *podSpec.TerminationGracePeriodSeconds < 0
+}
+
 // usesMultipleHugePageResources returns true if the pod spec uses more than
 // one size of hugepage
 func usesMultipleHugePageResources(podSpec *api.PodSpec) bool {
@@ -345,6 +349,8 @@ func GetValidationOptionsFromPodSpec(podSpec, oldPodSpec *api.PodSpec) apivalida
 	if oldPodSpec == nil {
 		return opts
 	}
+	// if old spec had negative terminationGracePeriodSeconds, we must allow it
+	opts.AllowNegativeTerminationGracePeriodSeconds = hasNegativeTerminationGracePeriodSeconds(oldPodSpec)
 	// if old spec used multiple huge page sizes, we must allow it
 	opts.AllowMultipleHugePageResources = opts.AllowMultipleHugePageResources || usesMultipleHugePageResources(oldPodSpec)
 	// if old spec used hugepages in downward api, we must allow it
diff --git a/pkg/apis/core/validation/validation.go b/pkg/apis/core/validation/validation.go
index 8f6ce285ce5..68e540a5cad 100644
--- a/pkg/apis/core/validation/validation.go
+++ b/pkg/apis/core/validation/validation.go
@@ -3185,6 +3185,8 @@ type PodValidationOptions struct {
 	AllowMultipleHugePageResources bool
 	// Allow pod spec to use hugepages in downward API
 	AllowDownwardAPIHugePages bool
+	// Allow negative termination grace period seconds
+	AllowNegativeTerminationGracePeriodSeconds bool
 }
 
 // ValidatePodSingleHugePageResources checks if there are multiple huge
@@ -3327,7 +3329,7 @@ func ValidatePodSpec(spec *core.PodSpec, podMeta *metav1.ObjectMeta, fldPath *fi
 		}
 	}
 
-	if spec.TerminationGracePeriodSeconds != nil {
+	if !opts.AllowNegativeTerminationGracePeriodSeconds && spec.TerminationGracePeriodSeconds != nil {
 		value := *spec.TerminationGracePeriodSeconds
 		if value < 0 || value > math.MaxInt32 {
 			allErrs = append(allErrs, field.Invalid(fldPath.Child("terminationGracePeriodSeconds"), value, validation.InclusiveRangeError(0, math.MaxInt32)))
```